python/cudf/udf_cpp/groupby/function.cu

```diff
@@ -0,0 +1,335 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/detail/utilities/device_operators.cuh>
+
+#include <cuda/atomic>
+
+#include <cooperative_groups.h>
+
+#include <limits>
+#include <type_traits>
+
+template <typename T>
+__device__ bool are_all_nans(cooperative_groups::thread_block const& block,
+                             T const* data,
+                             int64_t size)
+{
+  __shared__ bool result;
+
+  if (block.thread_rank() == 0) { result = true; }
+  block.sync();
+
+  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
+    if (not std::isnan(data[idx])) {
+      result = false;
+      break;
+    }
+  }
+
+  block.sync();
+  return result;
+}
+
+template <typename T>
+__device__ void device_sum(cooperative_groups::thread_block const& block,
+                           T const* data,
+                           int64_t size,
+                           T* sum)
+{
+  T local_sum = 0;
+
+  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
+    local_sum += data[idx];
+  }
+
+  cuda::atomic_ref<T, cuda::thread_scope_block> ref{*sum};
+  ref.fetch_add(local_sum, cuda::std::memory_order_relaxed);
+
+  block.sync();
+}
+
+template <typename T>
+__device__ void device_var(cooperative_groups::thread_block const& block,
+                           T const* data,
+                           int64_t size,
+                           double* var)
+{
+  T local_sum      = 0;
+  double local_var = 0;
+
+  __shared__ T block_sum;
+  if (block.thread_rank() == 0) { block_sum = 0; }
+  block.sync();
+
+  device_sum<T>(block, data, size, &block_sum);
+
+  auto const mean = static_cast<double>(block_sum) / static_cast<double>(size);
+
+  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
+    auto const delta = static_cast<double>(data[idx]) - mean;
+    local_var += delta * delta;
+  }
+
+  cuda::atomic_ref<double, cuda::thread_scope_block> ref{*var};
+  ref.fetch_add(local_var, cuda::std::memory_order_relaxed);
+  block.sync();
+
+  if (block.thread_rank() == 0) { *var = *var / static_cast<double>(size - 1); }
+  block.sync();
+}
+
+template <typename T>
+__device__ T BlockSum(T const* data, int64_t size)
+{
+  auto block = cooperative_groups::this_thread_block();
+
+  if constexpr (std::is_floating_point_v<T>) {
+    if (are_all_nans(block, data, size)) { return 0; }
+  }
+
+  __shared__ T block_sum;
+  if (block.thread_rank() == 0) { block_sum = 0; }
+  block.sync();
```

Comment:
Just a clarification for me. Is this `sync()` ok with a possible `return` at line 101?
That is, if one of threads in this block returns at line 101 then this `sync()` could be waiting forever.
Perhaps the `cooperative_groups` thread block handle this for us?

Reply:
The whole block is passed to `are_all_nans` and synced internally thus if it returns, all threads in the block will return. This is also the reason that I rewrote all the original functions by passing thread block as a parameter. This could largely reduce the chance that we unintentionally introduce the synchronization issues as you mentioned above.